plugin/role_set_utils.go

```diff
@@ -0,0 +1,161 @@
+
+	accountId *gcputil.ServiceAccountId
+	bindings  ResourceBindings
+	tokenGen  *TokenGenerator
+}
+
+func (rs *RoleSet) boundResources() *roleSetResources {
+	if rs.AccountId == nil {
+		return nil
+	}
+	return &roleSetResources{
+		accountId:   rs.AccountId,
+		rolesetName: rs.Name,
+		bindings:    rs.Bindings,
+		tokenGen:    rs.TokenGen,
+	}
+}
+
+// addWalsForRoleSetResources creates WALs to clean up a roleset's service account, bindings, and a key if needed.
+func (b *backend) addWalsForRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) error {
+	if boundResources == nil {
+		b.Logger().Debug("skip WALs for nil roleset resources")
+		return nil
+	}
+
+	_, err := framework.PutWAL(ctx, req.Storage, walTypeAccount, &walAccount{
+		RoleSet: boundResources.rolesetName,
+		Id:      *boundResources.accountId,
+	})
+	if err != nil {
+		return errwrap.Wrapf("unable to create WAL entry to clean up service account: {{err}}", err)
+	}
+
+	for resName, roleSet := range boundResources.bindings {
+		_, err := framework.PutWAL(ctx, req.Storage, walTypeIamPolicy, &walIamPolicy{
+			RoleSet:   boundResources.rolesetName,
+			AccountId: *boundResources.accountId,
+			Resource:  resName,
+			Roles:     roleSet.ToSlice(),
+		})
+		if err != nil {
+			return errwrap.Wrapf("unable to create WAL entry to clean up service account bindings: {{err}}", err)
+		}
+	}
+
+	if boundResources.tokenGen != nil {
+		return b.addServiceAccountKeyWal(ctx, req, boundResources.rolesetName, boundResources.accountId, boundResources.tokenGen.KeyName)
+	}
+	return nil
+}
+
+// addServiceAccountKeyWal creates WAL to clean up a roleset service account key (for access tokens) if needed.
+func (b *backend) addServiceAccountKeyWal(ctx context.Context, req *logical.Request, rolesetName string, accountId *gcputil.ServiceAccountId, keyName string) error {
+	if accountId == nil {
+		b.Logger().Debug("skip adding service account key WAL for nil account")
+		return nil
+	}
+
+	_, err := framework.PutWAL(ctx, req.Storage, walTypeAccount, &walAccountKey{
+		RoleSet:            rolesetName,
+		ServiceAccountName: accountId.ResourceName(),
+		KeyName:            keyName,
+	})
+	if err != nil {
+		return errwrap.Wrapf("unable to create WAL entry to clean up service account key: {{err}}", err)
+	}
+	return nil
+}
+
+// tryDeleteRoleSetResources tries to delete GCP resources previously managed by a roleset.
+// This assumes that deletion of these resources will already be guaranteed by WALs and will return errors
+// as a list of warnings instead.
+func (b *backend) tryDeleteRoleSetResources(ctx context.Context, req *logical.Request, boundResources *roleSetResources) []string {
+	if boundResources == nil {
+		b.Logger().Debug("skip deletion for nil roleset resources")
+		return nil
+	}
+
+	httpC, err := b.HTTPClient(req.Storage)
+	if err != nil {
+		return []string{err.Error()}
+	}
+
+	iamAdmin, err := iam.NewService(ctx, option.WithHTTPClient(httpC))
```

Comment:
You should be able to call [`b.IAMAdminClient()` without needing the http client](https://github.com/hashicorp/vault-plugin-secrets-gcp/blob/9c1dd008b73a781f4954215084a1b3d104d88a81/plugin/backend.go#L91)

Reply:
I'll switch it so it uses the cached version, I think I used this because we needed the HTTP client anyways for the iamHandle. 

Side question - is HTTP client is the biggest thing we wanted to cache?

Reply:
It's Russian doll cached. So IAMClient creates and caches an HTTP client too.